File: adapters/github-adapter/sdk/src/webhooks.ts
import { createHmac } from "crypto";
import { WebhookPayload, WebhookEvent } from "./types";

export function verifyWebhookSignature(
  payload: string | Buffer,
  signature: string,
  secret: string
): boolean {
  const expected = createHmac("sha256", secret)
    .update(payload)
    .digest("hex");
  const prefixed = `sha256=${expected}`;

  if (signature.length !== prefixed.length) return false;

  let mismatch = 0;
  for (let i = 0; i < prefixed.length; i++) {
    mismatch |= prefixed.charCodeAt(i) ^ signature.charCodeAt(i);
  }
  return mismatch === 0;
}

export function parseWebhookPayload(body: string): WebhookPayload {
  return JSON.parse(body) as WebhookPayload;
}

export type WebhookHandler = (payload: WebhookPayload) => void | Promise<void>;

export class WebhookRouter {
  private handlers = new Map<WebhookEvent | "*", WebhookHandler[]>();

  on(event: WebhookEvent | "*", handler: WebhookHandler): this {
    const list = this.handlers.get(event) ?? [];
    list.push(handler);
    this.handlers.set(event, list);
    return this;
  }

  async handle(payload: WebhookPayload): Promise<void> {
    const specific = this.handlers.get(payload.event) ?? [];
    const wildcard = this.handlers.get("*") ?? [];
    for (const handler of [...specific, ...wildcard]) {
      await handler(payload);
    }
  }
}

File: adapters/github-adapter/python-sdk/agentrail/webhooks.py
from __future__ import annotations

import hashlib
import hmac
import json
from collections import defaultdict
from typing import Awaitable, Callable

from .models import WebhookEvent, WebhookPayload


def verify_webhook_signature(payload: str | bytes, signature: str, secret: str) -> bool:
    if isinstance(payload, str):
        payload = payload.encode()
    expected = "sha256=" + hmac.new(
        secret.encode(), payload, hashlib.sha256
    ).hexdigest()
    return hmac.compare_digest(expected, signature)


def parse_webhook_payload(body: str) -> WebhookPayload:
    return WebhookPayload.model_validate(json.loads(body))


WebhookHandler = Callable[[WebhookPayload], Awaitable[None] | None]


class WebhookRouter:
    def __init__(self) -> None:
        self._handlers: dict[str, list[WebhookHandler]] = defaultdict(list)

    def on(self, event: WebhookEvent | str, handler: WebhookHandler) -> WebhookRouter:
        key = event.value if isinstance(event, WebhookEvent) else event
        self._handlers[key].append(handler)
        return self

    async def handle(self, payload: WebhookPayload) -> None:
        event_key = payload.event.value if isinstance(payload.event, WebhookEvent) else payload.event
        handlers = list(self._handlers.get(event_key, []))
        handlers.extend(self._handlers.get("*", []))
        for handler in handlers:
            result = handler(payload)
            if result is not None:
                await result
